plugins/keyboard-navigation/src/navigation.js

```diff
@@ -380,4 +380,505 @@ export class Navigation {
+   * @package
+   */
+  focusWorkspace(workspace) {
+    Blockly.hideChaff();
+    const reset = !!workspace.getToolbox();
+
+    this.resetFlyout(workspace, reset);
+    this.setState(workspace, Constants.STATE.WORKSPACE);
+    this.setCursorOnWorkspaceFocus(workspace);
+  }
+
+  /**
+   * Gets the cursor on the flyout's workspace.
+   * @param {!Blockly.WorkspaceSvg} workspace The main workspace the flyout is
+   *     on.
+   * @return {?Blockly.FlyoutCursor} The flyout's cursor or null if no flyout
+   *     exists.
+   * @protected
+   */
+  getFlyoutCursor(workspace) {
+    const flyout = workspace.getFlyout();
+    const cursor = flyout ? flyout.getWorkspace().getCursor() : null;
+
+    return /** @type {Blockly.FlyoutCursor} */ (cursor);
+  }
+
+  /**
+   * Hides the flyout cursor and optionally hides the flyout.
+   * @param {!Blockly.WorkspaceSvg} workspace The workspace.
+   * @param {boolean} shouldHide True if the flyout should be hidden.
+   * @protected
+   */
+  resetFlyout(workspace, shouldHide) {
+    if (this.getFlyoutCursor(workspace)) {
+      this.getFlyoutCursor(workspace).hide();
+      if (shouldHide) {
+        workspace.getFlyout().hide();
+      }
+    }
+  }
+
+  /**
+   * Tries to connect the given marker and cursor node.
+   * @param {!Blockly.WorkspaceSvg} workspace The main workspace.
+   * @param {!Blockly.ASTNode} markerNode The node to try to connect to.
+   * @param {!Blockly.ASTNode} cursorNode The node to connect to the markerNode.
+   * @return {boolean} True if the key was handled; false if something went
+   *     wrong.
+   * @protected
+   */
+  modify(workspace, markerNode, cursorNode) {
+    if (!this.modifyWarn(markerNode, cursorNode)) {
+      return false;
+    }
+
+    const markerType = markerNode.getType();
+    const cursorType = cursorNode.getType();
+
+    const cursorLoc = cursorNode.getLocation();
+    const markerLoc = markerNode.getLocation();
+    if (markerNode.isConnection() && cursorNode.isConnection()) {
+      const cursorConnection =
+          /** @type {!Blockly.RenderedConnection} */ (cursorLoc);
+      const markerConnection =
+          /** @type {!Blockly.RenderedConnection} */ (markerLoc);
+      return this.connect(cursorConnection, markerConnection);
+    } else if (
+        markerNode.isConnection() &&
+        (cursorType == Blockly.ASTNode.types.BLOCK ||
+         cursorType == Blockly.ASTNode.types.STACK)) {
+      const cursorBlock = /** @type {!Blockly.BlockSvg} */ (cursorLoc);
+      const markerConnection =
+          /** @type {!Blockly.RenderedConnection} */ (markerLoc);
+      return this.insertBlock(cursorBlock, markerConnection);
+    } else if (markerType == Blockly.ASTNode.types.WORKSPACE) {
+      const block = cursorNode ? cursorNode.getSourceBlock() : null;
+      return this.moveBlockToWorkspace(
+          /** @type {Blockly.BlockSvg} */ (block), markerNode);
+    }
+    this.warn('Unexpected state in setWorkspaceState.');
```

Comment:
this comment did not help my naming confusion because there is no function here called setWorkspaceState. either that or I am more confused than I realized.

Reply:
Yeah this must have been a copy and paste error. Fixed.